api/eds.proto

```diff
@@ -0,0 +1,183 @@
+  //    LoadBalanceRequest. This message contains per cluster/per Locality
+  //    information.
+  // 4. For each cluster, Envoy load balances incoming traffic to upstream hosts
+  //    based on per-zone weights and/or per-instance weights (if specified)
+  //    based on intra-zone LbPolicy.
+  // 5. When upstream hosts reply, they optionally add header <define header
+  //    name> with ASCII representation of EndpointLoadMetricStats.
+  // 6. Envoy aggregates load reports over the period of time given to it in
+  //    LoadBalanceResponse.load_reporting_interval. This includes aggregation
+  //    stats Envoy maintains by itself (total_requests, rpc_errors etc.) as
+  //    well as load metrics from upstream hosts.
+  // 7. When the timer of load_reporting_interval expires, Envoy sends new
+  //    LoadBalanceRequest filled with load reports for each cluster.
+  // 8. The management server uses the load reports from all reported Envoys
+  //    from around the world, computes global assignment and prepares traffic
+  //    assignment destined for each zone Envoys are located in. Goto 3.
+  // TODO(htuch): Add @amb67's diagram.
+  rpc BalanceLoad(stream LoadBalanceRequest)
+      returns (stream LoadBalanceResponse) {
+  }
+}
+
+message EndpointDiscoveryRequest {
+  repeated string cluster_name = 1;
+  Node node = 2;
+}
+
+enum HealthStatus {
+  UNKNOWN = 0;
+  HEALTHY = 1;
+  UNHEALTHY = 2;
+  // Connection draining in progress -
+  // https://aws.amazon.com/blogs/aws/elb-connection-draining-remove-instances-from-service-with-care/
+  // and
+  // https://cloud.google.com/compute/docs/load-balancing/enabling-connection-draining.
+  DRAINING = 3;
+  // This value is used by HDS Remote server. From Envoy’s perspective
+  // TIMEOUT = UNHEALTHY in case EDS returns HealthStatus.
+  TIMEOUT = 4;
+}
+
+message LbEndpoint {
+  // Endpoint host address.
+  Address address = 1;
+  // Optional health status when known and supplied by EDS server.
+  HealthStatus health_status = 2;
+  // The optional canary status of the upstream host. Envoy uses the canary
+  // status for various statistics and load balancing tasks documented
+  // elsewhere.
+  google.protobuf.BoolValue canary = 3;
+  // The optional load balancing weight of the upstream host, in the range 1 -
+  // 100. Envoy uses the load balancing weight in some of the built in load
+  // balancers.
+  // TODO(htuch): How does this combine with the LocalityEndpoints weight?
+  google.protobuf.UInt32Value load_balancing_weight = 4;
+}
+
+// All endpoints belonging to a Locality.
+message LocalityEndpoints {
+  Locality locality = 1;
+  repeated LbEndpoint lb_endpoints = 2;
+
+  // Optional: Per region/zone/sub_zone weight - range 1-100.
+  // TODO(htuch): Is there some restriction that all weights in a cluster must
+  // sum to 100? Do we want to be more specific and have per-zone and
+  // per-subzone as distinct weights for example? Can the weight also be 0?
```

Comment:
if remote server decides that weight is 0 and we should not send traffic to this upstream locality, it will not be part of EDS response